File: impl/go/internal/tool/protocol.go
// JSON-block tool-call protocol (Go parity of tool/protocol.ts).
// See docs/harness-design.md §3, §7. The LLM emits a fenced ```tool_calls```
// block; we parse it (with a tolerance ladder for slightly malformed JSON) and
// render results back as a ```tool_results``` block.
package tool

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

func nowNano() int64 { return time.Now().UnixNano() }

// ParsedTurn is the outcome of parsing one model response.
type ParsedTurn struct {
	Done       bool
	Summary    string
	Calls      []Call
	ParseError string // set when a block was present but unparseable
}

var blockRe = regexp.MustCompile("(?s)```tool_calls\\s*\\n(.*?)```")

// looseJSONRepair fixes common JSON mistakes before json.Unmarshal: block/line
// comments and trailing commas. It does not try to fix unbalanced quotes.
var (
	blockCommentRe = regexp.MustCompile(`(?s)/\*.*?\*/`)
	lineCommentRe  = regexp.MustCompile(`(^|\s)//[^\n]*`)
	trailingComma  = regexp.MustCompile(`,(\s*[}\]])`)
)

func looseJSONRepair(s string) string {
	out := blockCommentRe.ReplaceAllString(s, "")
	out = lineCommentRe.ReplaceAllString(out, "$1")
	out = trailingComma.ReplaceAllString(out, "$1")
	return strings.TrimSpace(out)
}

func tryParse(raw string) (map[string]any, error) {
	var v map[string]any
	if err := json.Unmarshal([]byte(raw), &v); err == nil {
		return v, nil
	}
	if err := json.Unmarshal([]byte(looseJSONRepair(raw)), &v); err != nil {
		return nil, err
	}
	return v, nil
}

// ParseToolCalls extracts and parses the tool_calls block from a model response.
func ParseToolCalls(resp string) ParsedTurn {
	m := blockRe.FindStringSubmatch(resp)
	if m == nil {
		return ParsedTurn{} // no block — director treats as idle
	}
	obj, err := tryParse(m[1])
	if err != nil {
		return ParsedTurn{
			ParseError: fmt.Sprintf("The tool_calls block was not valid JSON: %v. "+
				"Return exactly one ```tool_calls``` block whose body is a JSON object "+
				`with either {"calls":[...]} or {"done":true,"summary":"..."}.`, err),
		}
	}
	if d, ok := obj["done"].(bool); ok && d {
		summary, _ := obj["summary"].(string)
		return ParsedTurn{Done: true, Summary: summary}
	}
	rawCalls, _ := obj["calls"].([]any)
	var calls []Call
	for _, rc := range rawCalls {
		cm, ok := rc.(map[string]any)
		if !ok {
			continue
		}
		toolName, ok := cm["tool"].(string)
		if !ok {
			continue
		}
		args, _ := cm["args"].(map[string]any)
		if args == nil {
			args = map[string]any{}
		}
		calls = append(calls, Call{Tool: toolName, Args: args})
	}
	if len(calls) == 0 {
		return ParsedTurn{
			ParseError: `The tool_calls block had no usable calls. Each call needs a string "tool" ` +
				`and an "args" object, e.g. {"calls":[{"tool":"read","args":{"filePath":"src/Root.tsx"}}]}.`,
		}
	}
	return ParsedTurn{Calls: calls}
}

// RenderToolResults renders executed results as a ```tool_results``` block.
func RenderToolResults(results []CallResult) string {
	type wire struct {
		Tool       string `json:"tool"`
		OK         bool   `json:"ok"`
		Output     string `json:"output"`
		Truncated  bool   `json:"truncated,omitempty"`
		OutputPath string `json:"outputPath,omitempty"`
	}
	out := struct {
		Results []wire `json:"results"`
	}{}
	for _, r := range results {
		out.Results = append(out.Results, wire{r.Tool, r.OK, r.Output, r.Truncated, r.OutputPath})
	}
	b, _ := json.MarshalIndent(out, "", "  ")
	return "```tool_results\n" + string(b) + "\n```"
}

File: impl/ts/src/tool/protocol.ts
// JSON-block tool-call protocol — see docs/harness-design.md §3, §7.
// The LLM emits a fenced ```tool_calls``` block; we parse it (with a tolerance
// ladder, since models often produce slightly malformed JSON), and we render
// results back as a ```tool_results``` block.

import type { ToolCall, ToolCallResult } from "./types.js";

export interface ParsedTurn {
  // The model signalled it is finished.
  done: boolean;
  summary?: string;
  // Tool calls to execute this turn (empty when done or when no block found).
  calls: ToolCall[];
  // Set when a tool_calls block was present but could not be parsed — the
  // director feeds this back so the model can correct its formatting.
  parseError?: string;
}

const BLOCK_RE = /```tool_calls\s*\n([\s\S]*?)```/;

// Best-effort repair of common JSON mistakes before JSON.parse:
//   - trailing commas before } or ]
//   - // line comments and /* */ block comments
// We do NOT try to fix unbalanced quotes/braces — that way lies madness; we
// surface a clear error instead and let the model retry.
function looseJsonRepair(s: string): string {
  let out = s;
  // Strip block comments.
  out = out.replace(/\/\*[\s\S]*?\*\//g, "");
  // Strip line comments (but not inside strings — best effort: only when the
  // // is preceded by whitespace or line start, to avoid nuking URLs in values).
  out = out.replace(/(^|\s)\/\/[^\n]*/g, "$1");
  // Remove trailing commas.
  out = out.replace(/,(\s*[}\]])/g, "$1");
  return out.trim();
}

function tryParse(raw: string): { value?: unknown; error?: string } {
  try {
    return { value: JSON.parse(raw) };
  } catch {
    try {
      return { value: JSON.parse(looseJsonRepair(raw)) };
    } catch (e) {
      return { error: String((e as Error).message ?? e) };
    }
  }
}

// Extract and parse the tool_calls block from a model response.
export function parseToolCalls(resp: string): ParsedTurn {
  const m = resp.match(BLOCK_RE);
  if (!m) {
    // No block at all — the director treats this as "no progress" (it may nudge
    // the model or stop after an idle streak).
    return { done: false, calls: [] };
  }
  const { value, error } = tryParse(m[1]);
  if (error || typeof value !== "object" || value === null) {
    return {
      done: false,
      calls: [],
      parseError:
        `The tool_calls block was not valid JSON: ${error ?? "not an object"}. ` +
        `Return exactly one \`\`\`tool_calls\`\`\` block whose body is a JSON object ` +
        `with either {"calls":[...]} or {"done":true,"summary":"..."}.`,
    };
  }
  const obj = value as Record<string, unknown>;
  if (obj.done === true) {
    return {
      done: true,
      summary: typeof obj.summary === "string" ? obj.summary : "",
      calls: [],
    };
  }
  const rawCalls = Array.isArray(obj.calls) ? obj.calls : [];
  const calls: ToolCall[] = [];
  for (const c of rawCalls) {
    if (typeof c !== "object" || c === null) continue;
    const cc = c as Record<string, unknown>;
    if (typeof cc.tool !== "string") continue;
    const args = typeof cc.args === "object" && cc.args !== null ? (cc.args as Record<string, unknown>) : {};
    calls.push({ tool: cc.tool, args });
  }
  if (calls.length === 0) {
    return {
      done: false,
      calls: [],
      parseError:
        `The tool_calls block had no usable calls. Each call needs a string "tool" ` +
        `and an "args" object, e.g. {"calls":[{"tool":"read","args":{"filePath":"src/Root.tsx"}}]}.`,
    };
  }
  return { done: false, calls };
}

// Render executed results as a ```tool_results``` block to append to the
// conversation as the next user message.
export function renderToolResults(results: ToolCallResult[]): string {
  const body = {
    results: results.map((r) => ({
      tool: r.tool,
      ok: r.ok,
      output: r.output,
      ...(r.truncated ? { truncated: true } : {}),
      ...(r.outputPath ? { outputPath: r.outputPath } : {}),
    })),
  };
  return "```tool_results\n" + JSON.stringify(body, null, 2) + "\n```";
}
